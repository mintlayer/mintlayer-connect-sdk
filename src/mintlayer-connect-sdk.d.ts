interface MintlayerClient {
  isMintlayer: true;

  setNetwork: (net: 'testnet' | 'mainnet') => void;
  getNetwork: () => 'testnet' | 'mainnet';

  request: (options: { method: string; params?: Record<string, any> }) => Promise<any>;
  connect: () => Promise<string[]>;

  /** Returns the extended public key (xPub). Warning: Exposes all derived addresses. */
  getXPub: () => Promise<string>;

  getAddresses: () => Promise<string[]>;
  getBalance: () => Promise<string>;
  getDelegations: () => Promise<any[]>;

  buildTransaction: (options: {
    type?: 'Transfer' | 'CreateDelegationId' | 'IssueToken';
    params: Record<string, any>;
  }) => Promise<any>;

  transfer: (options: { to: string; amount: string }) => Promise<any>;
  delegate: (options: { poolId: string; amount?: string }) => Promise<any>;
  issueToken: (options: { tokenData: Record<string, any>; authority: string }) => Promise<any>;
  signTransaction: (tx: any) => Promise<any>;

  on: (eventName: string, callback: (data: any) => void) => void;
}

declare global {
  interface Window {
    mintlayer: MintlayerClient;
  }
}

export default MintlayerClient;
