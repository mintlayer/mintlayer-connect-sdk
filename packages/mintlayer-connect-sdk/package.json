{
  "name": "@mintlayer/mintlayer-connect-sdk",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "build": "mkdir -p dist && cp src/mintlayer-connect-sdk.js dist/",
    "serve": "npx http-server . -p 8080",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}
