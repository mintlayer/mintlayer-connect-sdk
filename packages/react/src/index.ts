export { MintlayerProvider, useMintlayer } from './context/MintlayerProvider';
