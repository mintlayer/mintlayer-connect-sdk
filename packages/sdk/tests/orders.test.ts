import { Client } from '../src/mintlayer-connect-sdk';
import fetchMock from 'jest-fetch-mock';

import { addresses, utxos } from './__mocks__/accounts/account_01'

let spy: jest.SpyInstance;

beforeEach(() => {
  fetchMock.resetMocks();

  spy = jest.spyOn(Client.prototype as any, 'buildTransaction');

  // эмуляция window.mojito
  (window as any).mojito = {
    isExtension: true,
    connect: jest.fn().mockResolvedValue(addresses),
    restore: jest.fn().mockResolvedValue(addresses),
    disconnect: jest.fn().mockResolvedValue(undefined),
    request: jest.fn().mockResolvedValue('signed-transaction'),
  };

  fetchMock.doMock();

  fetchMock.mockResponse(async req => {
    const url = req.url;

    if (url.endsWith('/chain/tip')) {
      return JSON.stringify({ height: 200000 });
    }

    if (url.includes('/token/')) {
      const tokenId = url.split('/token/').pop();
      if (tokenId === 'tmltk1jzgup986mh3x9n5024svm4wtuf2qp5vedlgy5632wah0pjffwhpqgsvmuq') {
        return JSON.stringify({
          "authority": "tmt1qyjlh9w9t7qwx7cawlqz6rqwapflsvm3dulgmxyx",
          "circulating_supply": {
            "atoms": "209000000000",
            "decimal": "2090"
          },
          "frozen": false,
          "is_locked": false,
          "is_token_freezable": true,
          "is_token_unfreezable": null,
          "metadata_uri": {
            "hex": "697066733a2f2f516d4578616d706c6548617368313233",
            "string": "ipfs://QmExampleHash123"
          },
          "next_nonce": 7,
          "number_of_decimals": 8,
          "token_ticker": {
            "hex": "58595a32",
            "string": "XYZ2"
          },
          "total_supply": {
            "Fixed": {
              "atoms": "100000000000000"
            }
          }
        });
      }
      if (tokenId === 'tmltk17jgtcm3gc8fne3su8s96gwj0yw8k2khx3fglfe8mz72jhygemgnqm57l7l') {
        return JSON.stringify({
          "authority": "tmt1qyjlh9w9t7qwx7cawlqz6rqwapflsvm3dulgmxyx",
          "circulating_supply": {
            "atoms": "209000000000",
            "decimal": "2090"
          },
          "frozen": false,
          "is_locked": false,
          "is_token_freezable": true,
          "is_token_unfreezable": null,
          "metadata_uri": {
            "hex": "697066733a2f2f516d4578616d706c6548617368313233",
            "string": "ipfs://QmExampleHash123"
          },
          "next_nonce": 7,
          "number_of_decimals": 11,
          "token_ticker": {
            "hex": "58595a32",
            "string": "XYZ2"
          },
          "total_supply": {
            "Fixed": {
              "atoms": "100000000000000"
            }
          }
        });
      }
      return JSON.stringify({ a: 'b' });
    }

    if(url.includes('/order/')) {
      const orderId = req.url.split('/order/').pop();

      if (orderId === 'tordr1thu5ykcdl0uj30g97wqkam7kart50lgzaq60edh8nq6zrn366lmql50gnu') {
        return {
          body: JSON.stringify({
            "ask_balance": {
              "atoms": "1000000000000",
              "decimal": "10"
            },
            "ask_currency": {
              "type": "Coin"
            },
            "conclude_destination": "tmt1qxf50ffxunjw557a9zf2et0vywkwjszyxyppa0py",
            "give_balance": {
              "atoms": "10000000000000",
              "decimal": "100"
            },
            "give_currency": {
              "token_id": "tmltk17jgtcm3gc8fne3su8s96gwj0yw8k2khx3fglfe8mz72jhygemgnqm57l7l",
              "type": "Token"
            },
            "initially_asked": {
              "atoms": "1000000000000",
              "decimal": "10"
            },
            "initially_given": {
              "atoms": "10000000000000",
              "decimal": "100"
            },
            "nonce": 0,
            "order_id": "tordr1thu5ykcdl0uj30g97wqkam7kart50lgzaq60edh8nq6zrn366lmql50gnu"
          }),
        };
      }
    }

    if(url.endsWith('/account')) {
      return {
        body: JSON.stringify({
          utxos: utxos,
        }),
      };
    }

    console.warn('No mock for:', url);
    return JSON.stringify({ error: 'No mock defined' });
  });
});

afterEach(() => {
  spy?.mockRestore();
});

test('create order - snapshot', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  await client.connect();

  await client.createOrder({
    ask_amount: 10,
    ask_token: 'Coin',
    give_amount: 100,
    give_token: 'tmltk1jzgup986mh3x9n5024svm4wtuf2qp5vedlgy5632wah0pjffwhpqgsvmuq',
    conclude_destination: 'tmt1q9874wgx6enm2mzfu0yxhzleu84pp00l95l7er5z',
  });

  const result = await spy.mock.results[0]?.value;

  expect(result).toMatchSnapshot();
});

test('fill order', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  await client.connect();

  await client.fillOrder({
    order_id: 'tordr1thu5ykcdl0uj30g97wqkam7kart50lgzaq60edh8nq6zrn366lmql50gnu',
    amount: 10,
    destination: 'tmt1q8kfn6vl835y4tj5yfsdvz7ay5cjnvhv952wftmt',
  });

  const result = await spy.mock.results[0]?.value;

  expect(result.JSONRepresentation).toStrictEqual({
    "inputs": [
      {
        "input": {
          "input_type": "AccountCommand",
          "command": "FillOrder",
          "order_id": "tordr1thu5ykcdl0uj30g97wqkam7kart50lgzaq60edh8nq6zrn366lmql50gnu",
          "fill_atoms": "1000000000000",
          "destination": "tmt1q8kfn6vl835y4tj5yfsdvz7ay5cjnvhv952wftmt",
          "nonce": "0"
        },
        "utxo": null
      },
      {
        "input": {
          "index": 1,
          "source_id": "af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab",
          "source_type": "Transaction",
          "input_type": "UTXO"
        },
        "utxo": {
          "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
          "type": "Transfer",
          "value": {
            "amount": {
              "atoms": "1703205604300000",
              "decimal": "17032.056043"
            },
            "type": "Coin"
          }
        }
      }
    ],
    "outputs": [
      {
        "type": "Transfer",
        "value": {
          "type": "TokenV1",
          "token_id": "tmltk17jgtcm3gc8fne3su8s96gwj0yw8k2khx3fglfe8mz72jhygemgnqm57l7l",
          "amount": {
            "atoms": "10000000000000",
            "decimal": "100"
          }
        },
        "destination": "tmt1q8kfn6vl835y4tj5yfsdvz7ay5cjnvhv952wftmt"
      },
      {
        "type": "Transfer",
        "value": {
          "type": "Coin",
          "amount": {
            "atoms": "1702005604300000",
            "decimal": "17020.056043"
          }
        },
        "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6"
      }
    ]
  })
});


test('conclude order - snapshot', async () => {
  const client = await Client.create({ network: 'testnet', autoRestore: false });

  fetchMock.mockResponseOnce(async req => {
    const url = req.url;

    if(url.includes('/order/')) {
      const orderId = req.url.split('/order/').pop();

      if (orderId === 'tordr1thu5ykcdl0uj30g97wqkam7kart50lgzaq60edh8nq6zrn366lmql50gnu') {
        return {
          body: JSON.stringify({
            "ask_balance": {
              "atoms": "100000000000",
              "decimal": "1"
            },
            "ask_currency": {
              "type": "Coin"
            },
            "conclude_destination": "tmt1qxf50ffxunjw557a9zf2et0vywkwjszyxyppa0py",
            "give_balance": {
              "atoms": "1000000000000",
              "decimal": "10"
            },
            "give_currency": {
              "token_id": "tmltk17jgtcm3gc8fne3su8s96gwj0yw8k2khx3fglfe8mz72jhygemgnqm57l7l",
              "type": "Token"
            },
            "initially_asked": {
              "atoms": "1000000000000",
              "decimal": "10"
            },
            "initially_given": {
              "atoms": "10000000000000",
              "decimal": "100"
            },
            "nonce": 1,
            "order_id": "tordr1thu5ykcdl0uj30g97wqkam7kart50lgzaq60edh8nq6zrn366lmql50gnu"
          }),
        };
      }
    }

    console.warn('No mock for:', url);
    return JSON.stringify({ error: 'No mock defined' });
  });

  await client.connect();

  await client.concludeOrder({
    order_id: 'tordr1thu5ykcdl0uj30g97wqkam7kart50lgzaq60edh8nq6zrn366lmql50gnu',
  });

  const result = await spy.mock.results[0]?.value;

  expect(result.JSONRepresentation).toStrictEqual({
    "inputs": [
      {
        "input": {
          "input_type": "AccountCommand",
          "command": "ConcludeOrder",
          "destination": "tmt1qxf50ffxunjw557a9zf2et0vywkwjszyxyppa0py",
          "order_id": "tordr1thu5ykcdl0uj30g97wqkam7kart50lgzaq60edh8nq6zrn366lmql50gnu",
          "nonce": 1
        },
        "utxo": null
      },
      {
        "input": {
          "index": 1,
          "source_id": "af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab",
          "source_type": "Transaction",
          "input_type": "UTXO"
        },
        "utxo": {
          "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
          "type": "Transfer",
          "value": {
            "amount": {
              "atoms": "1703205604300000",
              "decimal": "17032.056043"
            },
            "type": "Coin"
          }
        }
      }
    ],
    "outputs": [
      {
        "type": "Transfer",
        "destination": "tmt1qxf50ffxunjw557a9zf2et0vywkwjszyxyppa0py",
        "value": {
          "type": "Coin",
          "amount": {
            "decimal": "9",
            "atoms": "900000000000"
          }
        }
      },
      {
        "type": "Transfer",
        "destination": "tmt1qxf50ffxunjw557a9zf2et0vywkwjszyxyppa0py",
        "value": {
          "type": "TokenV1",
          "token_id": "tmltk17jgtcm3gc8fne3su8s96gwj0yw8k2khx3fglfe8mz72jhygemgnqm57l7l",
          "amount": {
            "decimal": "10",
            "atoms": "1000000000000"
          }
        }
      },
      {
        "type": "Transfer",
        "value": {
          "type": "Coin",
          "amount": {
            "atoms": "1703005604300000",
            "decimal": "17030.056043"
          }
        },
        "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6"
      }
    ]
  })

  // expect(result).toMatchSnapshot();
});
