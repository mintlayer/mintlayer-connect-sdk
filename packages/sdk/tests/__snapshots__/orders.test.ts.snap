// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create order - snapshot 1`] = `
{
  "BINRepresentation": {
    "inputs": [
      Uint8Array [
        0,
        0,
        175,
        59,
        95,
        173,
        32,
        246,
        249,
        126,
        178,
        16,
        147,
        78,
        148,
        33,
        118,
        247,
        247,
        208,
        247,
        4,
        35,
        89,
        6,
        89,
        238,
        14,
        2,
        23,
        5,
        58,
        124,
        171,
        1,
        0,
        0,
        0,
      ],
      Uint8Array [
        0,
        0,
        205,
        206,
        228,
        164,
        72,
        35,
        151,
        140,
        197,
        2,
        69,
        239,
        245,
        24,
        165,
        98,
        196,
        70,
        30,
        116,
        254,
        226,
        206,
        47,
        70,
        202,
        222,
        130,
        47,
        101,
        126,
        13,
        0,
        0,
        0,
        0,
      ],
    ],
    "outputs": [
      Uint8Array [
        11,
        1,
        79,
        234,
        185,
        6,
        214,
        103,
        181,
        108,
        73,
        227,
        200,
        107,
        139,
        249,
        225,
        234,
        16,
        189,
        255,
        45,
        0,
        7,
        0,
        16,
        165,
        212,
        232,
        2,
        144,
        145,
        192,
        148,
        250,
        221,
        226,
        98,
        206,
        143,
        85,
        96,
        205,
        213,
        203,
        226,
        84,
        0,
        209,
        153,
        111,
        208,
        74,
        106,
        42,
        119,
        110,
        240,
        201,
        41,
        117,
        194,
        7,
        0,
        228,
        11,
        84,
        2,
      ],
      Uint8Array [
        0,
        0,
        15,
        224,
        68,
        24,
        228,
        223,
        12,
        6,
        1,
        134,
        236,
        69,
        4,
        87,
        208,
        154,
        217,
        128,
        115,
        147,
        216,
        156,
        236,
        156,
        113,
        214,
        32,
        96,
        33,
      ],
      Uint8Array [
        0,
        2,
        144,
        145,
        192,
        148,
        250,
        221,
        226,
        98,
        206,
        143,
        85,
        96,
        205,
        213,
        203,
        226,
        84,
        0,
        209,
        153,
        111,
        208,
        74,
        106,
        42,
        119,
        110,
        240,
        201,
        41,
        117,
        194,
        7,
        0,
        4,
        107,
        244,
        20,
        1,
        134,
        236,
        69,
        4,
        87,
        208,
        154,
        217,
        128,
        115,
        147,
        216,
        156,
        236,
        156,
        113,
        214,
        32,
        96,
        33,
      ],
    ],
    "transactionsize": 447,
  },
  "HEXRepresentation_unsigned": "0100080000af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab010000000000cdcee4a44823978cc50245eff518a562c4461e74fee2ce2f46cade822f657e0d000000000c0b014feab906d667b56c49e3c86b8bf9e1ea10bdff2d00070010a5d4e8029091c094fadde262ce8f5560cdd5cbe25400d1996fd04a6a2a776ef0c92975c20700e40b540200000fe04418e4df0c060186ec450457d09ad9807393d89cec9c71d620602100029091c094fadde262ce8f5560cdd5cbe25400d1996fd04a6a2a776ef0c92975c20700046bf4140186ec450457d09ad9807393d89cec9c71d6206021",
  "JSONRepresentation": {
    "inputs": [
      {
        "input": {
          "index": 1,
          "input_type": "UTXO",
          "source_id": "af3b5fad20f6f97eb210934e942176f7f7d0f70423590659ee0e0217053a7cab",
          "source_type": "Transaction",
        },
        "utxo": {
          "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
          "type": "Transfer",
          "value": {
            "amount": {
              "atoms": "1703205604300000",
              "decimal": "17032.056043",
            },
            "type": "Coin",
          },
        },
      },
      {
        "input": {
          "index": 0,
          "input_type": "UTXO",
          "source_id": "cdcee4a44823978cc50245eff518a562c4461e74fee2ce2f46cade822f657e0d",
          "source_type": "Transaction",
        },
        "utxo": {
          "destination": "tmt1qyjlh9w9t7qwx7cawlqz6rqwapflsvm3dulgmxyx",
          "type": "Transfer",
          "value": {
            "amount": {
              "atoms": "100000000000",
              "decimal": "1000",
            },
            "token_id": "tmltk1jzgup986mh3x9n5024svm4wtuf2qp5vedlgy5632wah0pjffwhpqgsvmuq",
            "type": "TokenV1",
          },
        },
      },
    ],
    "outputs": [
      {
        "ask_balance": {
          "atoms": "1000000000000",
          "decimal": "10",
        },
        "ask_currency": {
          "type": "Coin",
        },
        "conclude_destination": "tmt1q9874wgx6enm2mzfu0yxhzleu84pp00l95l7er5z",
        "give_balance": {
          "atoms": "10000000000",
          "decimal": "100",
        },
        "give_currency": {
          "token_id": "tmltk1jzgup986mh3x9n5024svm4wtuf2qp5vedlgy5632wah0pjffwhpqgsvmuq",
          "type": "TokenV1",
        },
        "initially_asked": {
          "atoms": "1000000000000",
          "decimal": "10",
        },
        "initially_given": {
          "atoms": "10000000000",
          "decimal": "100",
        },
        "type": "CreateOrder",
      },
      {
        "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
        "type": "Transfer",
        "value": {
          "amount": {
            "atoms": "1703005604300000",
            "decimal": "17030.056043",
          },
          "type": "Coin",
        },
      },
      {
        "destination": "tmt1qxrwc3gy2lgf4kvqwwfa388vn3cavgrqyyrgswe6",
        "type": "Transfer",
        "value": {
          "amount": {
            "atoms": "90000000000",
            "decimal": "900",
          },
          "token_id": "tmltk1jzgup986mh3x9n5024svm4wtuf2qp5vedlgy5632wah0pjffwhpqgsvmuq",
          "type": "TokenV1",
        },
      },
    ],
  },
  "transaction_id": "e1b6ae17eb61672986728072d283baeb629ae6a19203609f3a9cace9c3ca4682",
}
`;
